doc/adr/0013-tx-validity-time-range-upper-bound.rst

```diff
@@ -0,0 +1,200 @@
+  > let txValidRange = toCardanoValidityRange r
+  > txValidRange
+  Right (TxValidityLowerBound ValidityLowerBoundInBabbageEra (SlotNo 1),TxValidityNoUpperBound ValidityNoUpperBoundInBabbageEra)
+
+When creating the ``TxInfo``, ``cardano-ledger`` will convert the previous ``cardano-api`` validity slot range to:
+
+.. code-block:: haskell
+
+  (Interval (LowerBound (Finite 1000) True) (UpperBound PosInf True))
+
+In practical reasoning, ``LowerBound (Finite 999) False`` and ``LowerBound (Finite 1000) True`` are
+equal considering the precision of 1000 milliseconds per slot.
+However, given ``Interval`` semantics, these are not the same values.
+Therefore, if the constraint ``mustValidateIn interval`` is used both to create a transaction and
+inside a Plutus script (corresponds to the check ``interval `contains` txInfoValidRange
+scriptContextTxInfo``), then the Plutus script will yield ``False``.
+
+We can identify a similar behavior with the upper bound.
+
+.. code-block:: haskell
+
+  > let sc = SlotConfig 1000 0
+  > let interval = (Interval (LowerBound NegInf True) (UpperBound (Finite 999) True))
+  > let r = posixTimeRangeToContainedSlotRange sc interval
+  > r
+  Interval {ivFrom = LowerBound NegInf True, ivTo = UpperBound (Finite (Slot {getSlot = 0})) True}
+  > let txValidRange = toCardanoValidityRange r
+  > txValidRange
+  Right (TxValidityNoLowerBound,TxValidityUpperBound ValidityUpperBoundInBabbageEra (SlotNo 1))
+
+When creating the ``TxInfo``, ``cardano-ledger`` will convert the previous ``cardano-api`` validity slot range to:
+
+.. code-block:: haskell
+
+  (Interval (LowerBound NegInf True) (UpperBound (Finite 1000) False))
+
+Again, a Plutus script with ``interval `contains` txInfoValidRange scriptContextTxInfo`` will yield ``False``.
+
+Additionnaly, the current behavior makes it hard to reason about how a ``POSIXTime`` gets translated
+into a ``Slot`` when creating a transaction.
+Ultimately, a DApp developer should have control over how his ``POSIXTime`` gets translated to a
+``Slot``.
+
+Decision
+--------
+
+* We will create the following datatype:
+
+  .. code-block:: haskell
+
+    -- | ValidityInterval is a half open interval. Closed (inclusive) on the bottom, open
+    -- (exclusive) on the top. A 'Nothing' on the bottom is negative infinity, and a 'Nothing'
+    -- on the top is positive infinity.
+    data ValidityInterval a = ValidityInterval
+      { invalidBefore :: !(Maybe a) -- ^ Inclusive lower bound or negative infinity
+      , invalidHereafter :: !(Maybe a) -- ^ Exclusive upper bound or positive infinity
+      }
+
+* We will add the following constraint and smart constructor:
+
+  .. code-block:: haskell
+
+    data TxConstraint =
+      ...
+      MustValidateInTimeRange !(ValidityInterval POSIXTime)
+
+    mustValidateInTimeRange :: !(ValidityInterval POSIXTime) -> TxConstraints
+
+* We will deprecate the ``MustValidateIn`` constraint and the ``mustValidateIn`` smart constructor
```

Comment:
Perhaps we should remove `MustValidateIn` immediately, since it would be dead weight in the script size.

Reply:
:+1: